contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol

```diff
@@ -163,27 +150,66 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
     return targetsToFund;
   }
 
-  /// @notice Send funds to the targets provided.
-  /// @param targetAddresses the list of targets to fund
   function topUp(address[] memory targetAddresses) public whenNotPaused {
-    uint256 topUpAmount = s_topUpAmount;
-    uint256 stopIdx = targetAddresses.length;
-    uint256 numCanFund = LINK_TOKEN.balanceOf(address(this)) / topUpAmount;
-    stopIdx = numCanFund < stopIdx ? numCanFund : stopIdx;
-    for (uint256 idx = 0; idx < stopIdx; idx++) {
-      (bool exists, uint256 minBalance) = s_watchList.tryGet(targetAddresses[idx]);
-      if (!exists) {
-        emit TopUpBlocked(targetAddresses[idx]);
-        continue;
+    MonitoredAddress memory target;
+    uint256 localBalance = i_linkToken.balanceOf(address(this));
+    address[] memory targetsTopUp = new address[](targetAddresses.length);
+    for (uint256 idx = 0; idx < targetAddresses.length; idx++) {
+      address targetAddress = targetAddresses[idx];
+      target = s_targets[targetAddress];
+      (bool needsFunding, ) = _needsFunding(targetAddress, target.minBalance);
+      if (target.isActive && localBalance >= target.topUpAmount && !(targetAddress == address(0)) && needsFunding) {
```

Comment:
this doesn't check the minWaitPeriod - is that intentional? Looks like we already check if targetAddress == address(0) in `_needsFunding()`. And then we do it twice again here - is that necessary?

Reply:
Yes, updated with the latest changes!